### PR TITLE
[query] Add Equal method to Query and simplify query matcher

### PR DESCRIPTION
**Pull request by jeromefroe**
This PR moves the logic for query equality into a `Equal` method on the `Query` interface. This makes it simpler to add new types of queries since we don't need to separately update the query matcher.

### CODE REVIEW

File: idx/query_matcher_test.go [hunk 1]
```diff
@@ -1,126 +0,0 @@
...
-
-import (
-	"testing"
-
-	"github.com/m3db/m3ninx/idx"
-
-	"github.com/stretchr/testify/require"
-)
-
-func TestQueryMatcherTermQuery(t *testing.T) {
```
[the remark this answers is not in the document]
Comment (jeromefroe) [reply idx/query_matcher_test.go]: Doesn't seem like the logic for matching should live here and they are redundant now anyway.

File: search/query/disjunction.go [hunk 2]
```diff
@@ -71,3 +71,23 @@ func (q *DisjuctionQuery) Searcher(rs index.Readers) (search.Searcher, error) {
...
+	inner, ok := o.(*DisjuctionQuery)
+	if !ok {
+		return false
+	}
+
+	if len(q.Queries) != len(inner.Queries) {
+		return false
+	}
+
+	// TODO: Should order matter?
```
Comment (prateek) [inline search/query/disjunction.go]: nah i think it's fine to not care about this

File: search/query/conjunction.go [hunk 3]
```diff
@@ -71,3 +71,23 @@ func (q *ConjuctionQuery) Searcher(rs index.Readers) (search.Searcher, error) {
...
+	inner, ok := o.(*ConjuctionQuery)
+	if !ok {
+		return false
+	}
+
+	if len(q.Queries) != len(inner.Queries) {
+		return false
+	}
+
+	// TODO: Should order matter?
```
Comment (prateek) [inline search/query/conjunction.go]: same here

File: idx/query_matcher_test.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (prateek) [reply idx/query_matcher_test.go]: re: logic for matching - sure but that doesn't have anything to do with the tests for the matcher, no?

[the remark this answers is not in the document]
Comment (jeromefroe) [reply idx/query_matcher_test.go]: Seems like we're leaking that logic here, ideally the matcher should just call out to the `Equal` method.

File: search/query/disjunction.go (hunk 2 above)
Comment (jeromefroe) [reply search/query/disjunction.go]: To not care about order? We'll need to sort the queries in that case

File: idx/query_matcher_test.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (prateek) [reply idx/query_matcher_test.go]: talked offline - we'll make sure the `Equal()` behaviour is identical and then kill these tests

**Comment by jeromefroe** [comment]
@prateek mind taking another look when you get a chance, updated the equality methods for conjunction and disjunction to match the previous behavior

**Pull request merged**